IOS/USB: Emulate Wii Speak using cubeb

sepalani opened the pull request:
This PR is based on https://github.com/dolphin-emu/dolphin/pull/12567. 

I don't have much knowledge regarding the technical details of the implementation. I mainly fixed the Windows support, the heisenbug due to uninitialised members and finalised the config integration.


Ready to be reviewed and tested.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dolphin-emu/dolphin/12769)
<!-- Reviewable:end -->

sepalani commented:
I fixed a few more things. The Wii Speak Channel appears to work properly now _(its initial test won't work though, as we don't emulate the echo reduction)_. I also added the Android support.

To test this PR, you'll need to enable the emulated Wii Speak under `Tools > Emulated USB devices > Wii Speak` and to check the two checkboxes to enable the emulation and to "connect" it. On Android, it's located under `Settings > Config > Wii > Emulated USB Devices` and you'll also need to enable the microphone permission for the app.

The Wii Speak Channel can be tested using a regular one _(but by redirecting Nintendo domains to nothing like localhost, resulting in a skippable 20100 error)_ or via a Riiconnect24 patched version _(AFAICT, [WiiLink24 patcher](https://www.wiilink24.com/guide/install) doesn't seem to offer it currently, [Wiimmfi WAD Patcher](https://wiimmfi.de/patcher/wiiware) does)_. A Wii friend is required in order to be able to record Wii messages like this: https://www.youtube.com/watch?v=4kpNID36pW0

This PR can be tested on Monster Hunter 3 as well. You will need to patch the game to go online using alternative servers. You will need a friend in-game also using a Wii Speak _(Dolphin or a real Wii)_ to be able to speak in a city.

I haven't checked this PR with other games yet, such as Animal Crossing. I'll when I'll get the time to do so. Otherwise, this PR is still ready to be reviewed & tested.

Florin9doi commented on Source/Core/Core/IOS/USB/Emulated/WiiSpeak.h:
```diff
@@ -0,0 +1,102 @@
+  bool m_device_attached = false;
+  bool init = false;
+  std::unique_ptr<Microphone> m_microphone{};
+  const DeviceDescriptor m_device_descriptor{0x12,  0x1,    0x200,  0,   0,   0,   0x10,
```
```suggestion
  const DeviceDescriptor m_device_descriptor{
    .bLength = 0x12,
    .bDescriptorType = 0x01,
    .bcdUSB = 0x0200,
    .bDeviceClass = 0x00,
   ...
```

sepalani replied:
AFAICT, this syntax isn't used in the current codebase for similar device classes. I wouldn't mind changing it in a follow up PR.

JosJuice commented on Source/Android/app/src/main/AndroidManifest.xml:
```diff
@@ -28,6 +28,7 @@
     <uses-permission
         android:name="android.permission.VIBRATE"
         android:required="false"/>
+    <uses-permission android:name="android.permission.RECORD_AUDIO"/>
```
```suggestion
    <uses-permission
        android:name="android.permission.RECORD_AUDIO"
        android:required="false"/>
```

Otherwise Google Play can prevent Dolphin from being installed on devices with no microphone.

sepalani replied:
AFAICT, cubeb doesn't prompt anything and just [fails to initialise](https://github.com/mozilla/cubeb/blob/19fcbefe1a9c5e22f8111af251df27b41658bc77/src/cubeb_aaudio.cpp#L987-L994).

Should we check this permission and ask the user the appropriate permission? If so, should it be done in the C++ code before the initialisation or in the Android code when using the settings toggle?

JosJuice replied:
Sorry, I missed your comments until you pushed a new commit now.

Yes, I think we should. It's probably best to do it both when turning on the setting and when the C++ code tries to use it. The former because that makes the purpose of the permission the most clear to the user, and the latter in case the permission has been revoked by the user (or automatically by the OS because the app hasn't been used for a while).

JosJuice commented on Source/Android/app/src/main/res/values/strings.xml:
```diff
@@ -923,4 +923,6 @@ It can efficiently compress both junk data and encrypted Wii data.
     <string name="select_compatible_figure">Please select a compatible figure file</string>
 
+    <string name="emulate_wii_speak">Wii Speak</string>
+    <string name="disconnect_wii_speak">Mute Wii Speak</string>
```
"Mute" doesn't match what the setting is called in the code or in DolphinQt. Is this something that was overlooked, or is it intentional?

sepalani replied:
It is intended since it is presented differently on PC and Android. On PC it has its own dedicated window (similar to the emulated Infinity base):
![image](https://github.com/dolphin-emu/dolphin/assets/7890055/a72f3d1b-b31d-4413-b056-fd10c9ef7686)

Whereas on Android, it's only a toggle settings below the one enabling the Wii Speak. So to make it clearer for the end-user I decided to rename it (as the effect and purpose of disconnecting it is not to record audio). I feel that having a "Wii Speak" toggle, plus a "Disconnect Wii Speak" toggle right under it sounds strange from a UI point of view.

I'm open to better ways of presenting this option to the user.

JosJuice replied:
What does the setting do from the perspective of the emulated software? Does it disconnect the Wii Speak USB device, or does it just make the audio be silence? If it's the latter, I think both the Android UI and the PC UI are okay, but the internal name for the setting would make more sense as "mute" than "connect". "Connect" makes sense in the PC UI only because it's in the context of whether you want to connect a host microphone.

JosJuice reviewed (commented):
How does the permission prompt work on Android? Does cubeb automagically make the permission prompt show up?

sepalani commented:
@JosJuice 
> What does the setting do from the perspective of the emulated software? Does it disconnect the Wii Speak USB device, or does it just make the audio be silence? If it's the latter, I think both the Android UI and the PC UI are okay, but the internal name for the setting would make more sense as "mute" than "connect". "Connect" makes sense in the PC UI only because it's in the context of whether you want to connect a host microphone.

Ideally, it should have been used to disconnect the device. However, I couldn't find a way to do so and it had a side effect so I chose not to process the microphone data instead.
https://github.com/dolphin-emu/dolphin/blob/35436ac19dec5d31701f1a4432727a3feacc633c/Source/Core/Core/IOS/USB/Emulated/WiiSpeak.cpp#L97-L101
https://github.com/dolphin-emu/dolphin/blob/35436ac19dec5d31701f1a4432727a3feacc633c/Source/Core/Core/IOS/USB/Emulated/Microphone.cpp#L252-L255

> Yes, I think we should. It's probably best to do it both when turning on the setting and when the C++ code tries to use it. The former because that makes the purpose of the permission the most clear to the user, and the latter in case the permission has been revoked by the user (or automatically by the OS because the app hasn't been used for a while).

Apparently, enabling the permission after cubeb was initialised doesn't seem to work. I added a warning message before requesting the permission to warn the user about this issue. This part (especially the android code) needs to be properly reviewed as it's been a while since I've programmed for Android and I'm unsure if it's the proper way to do it.

JosJuice commented on Source/Android/app/src/main/res/values/strings.xml:
```diff
@@ -923,4 +923,8 @@ It can efficiently compress both junk data and encrypted Wii data.
 
+    <string name="emulate_wii_speak">Wii Speak</string>
+    <string name="disconnect_wii_speak">Mute Wii Speak</string>
+    <string name="wii_speak_permission_warning">Missing RECORD_AUDIO permission</string>
+    <string name="wii_speak_permission_warning_description">Wii Speak emulation requires RECORD_AUDIO permission. Please enable this permission to allow your microphone use in-game. You might need to restart the game for the permission to be effective. </string>
```
```suggestion
    <string name="wii_speak_permission_warning">Missing Microphone Permission</string>
    <string name="wii_speak_permission_warning_description">Wii Speak emulation requires microphone permission. Please enable this permission to allow your microphone use in-game. You might need to restart the game for the permission to be effective.</string>
```

Also, it would be nice UX to not mention having to restart if no game is running. (You can check this using NativeLibrary.IsRunning.)

JosJuice replied:
Please also note that displayAlertMsg uses a toast message if we're not currently in EmulationActivity. Toast messages are very limited in how much text they display on modern versions of Android, and for good reason – they only stick on screen for a short amount of time. If you want to display this message as a toast, it should be as short as possible. At the very least you should get rid of the "Please enable this permission to allow your microphone use in-game." sentence.

sepalani replied:
The warning shouldn't be displayed when the game isn't running as the change is done by the settings switch and the activity is passed to the function. The C++ code should only be triggered when the emulation is running and when the Microphone class wants to start a stream, so I don't think using NativeLibrary.IsRunning is necessary here.

Otherwise, I shortened the message as you suggested.

JosJuice replied:
Sorry, I had missed the `activity == null` check in `requestRecordAudioPermission`. Okay, that sounds reasonable then. But consider expanding the comment in `requestRecordAudioPermission` to mention why we need the warning when calling from C++ specifically.

sepalani replied:
Done.

JMC47 commented:
Is anyone else planning to review this?  Is this considered finished at this point?

JMC47 commented:
Does anyone else want to review this?  I can give this a test on Android, but I won't be reviewing the code.

MayImilae commented:
Are you still intending to add a microphone selector? Having a window with a single toggle is quite odd, as in that case you could just put the toggle itself into the file menu.

sepalani commented:
> Are you still intending to add a microphone selector? Having a window with a single toggle is quite odd, as in that case you could just put the toggle itself into the file menu.

Doesn't the selector appear when the emulated device is enabled? I'm just using a layout similar to the one of the Infinity Base, another emulated USB device. I'm open to suggestion regarding this window design. The only constraint that it has to take into account _(currently)_ is that it cannot enable or disable the emulated USB device when the emulation has started _(i.e. it has to be chosen before launching the game)_.

![image](https://github.com/user-attachments/assets/c41b6586-3af3-40dd-94c3-6e5b9b62a912)
![image](https://github.com/user-attachments/assets/d0698ab5-1f92-4aa9-ae8a-65ad13183826)
![image](https://github.com/user-attachments/assets/6bf1f500-e1e5-41d1-b71a-2631fc9d96f3)

MayImilae commented:
When I tested this I only saw what was in the top window. What scenario does that occur?

JosJuice commented on Source/Core/Core/Config/MainSettings.cpp:
```diff
@@ -572,6 +572,15 @@ const Info<bool> MAIN_EMULATE_SKYLANDER_PORTAL{
+    {System::Main, "EmulatedUSBDevices", "WiiSpeakMicrophone"}, ""};
+
+const Info<bool> MAIN_WII_SPEAK_CONNECTED{{System::Main, "EmulatedUSBDevices", "WiiSpeakConnected"},
+                                          false};
```
I think the source code name of this setting is confusing, since the obvious interpretation is that it controls whether the Wii Speak is connected to the console, which isn't actually the case, as it controls whether a *host* microphone is connected. My suggestion for a clearer name would be `MAIN_WII_SPEAK_MUTED` (with an inverted meaning), but I'm open to other options too (maybe something that makes it more obviously connected to `MAIN_WII_SPEAK_MICROPHONE`?).

JosJuice commented on Source/Core/Core/IOS/USB/Emulated/Microphone.cpp:
```diff
@@ -0,0 +1,261 @@
+#ifdef _WIN32
+  });
+  sync_event.Wait();
+#endif
```
This pattern occurs several times in this file. I recommend creating a helper function that handles the boilerplate and lets you pass in a lambda.

JosJuice commented on Source/Core/Core/IOS/USB/Emulated/WiiSpeak.cpp:
```diff
@@ -0,0 +1,412 @@
+  {
+  case USBHDR(DIR_DEVICE2HOST, TYPE_STANDARD, REC_INTERFACE, REQUEST_GET_INTERFACE):
+  {
+    const std::array<u8, 1> data{1};
```
Personally I would use a simple `u8` here and then use the `&` operator in place of `.data()`. But I suppose an `std::array` is fine too if you prefer it.

JosJuice commented on Source/Core/Core/IOS/USB/Emulated/WiiSpeak.h:
```diff
@@ -0,0 +1,102 @@
+
+  EmulationKernel& m_ios;
+  const u16 m_vid = 0x57E;
+  const u16 m_pid = 0x0308;
```
Leading zero for one but not the other?

JosJuice commented on Source/Core/DolphinQt/EmulatedUSB/WiiSpeakWindow.cpp:
```diff
@@ -0,0 +1,110 @@
+  checkbox_group->setLayout(checkbox_layout);
+  main_layout->addWidget(checkbox_group);
+
+  m_config_group = new QGroupBox(tr("Microphone configuration"));
```
```suggestion
  m_config_group = new QGroupBox(tr("Microphone Configuration"));
```

JosJuice commented on Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/features/settings/ui/viewholder/SwitchSettingViewHolder.kt:
```diff
@@ -57,6 +59,13 @@ class SwitchSettingViewHolder(
 
+            if (setting.setting === BooleanSetting.MAIN_EMULATE_WII_SPEAK && isChecked) {
+                if (!PermissionsHandler.hasRecordAudioPermission(itemView.context)) {
+                  val settingsActivity = itemView.context as Activity
```
This cast feels a bit fragile. How about: `adapter.fragmentView.fragmentActivity`

JosJuice commented on Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/utils/ActivityTracker.kt:
```diff
@@ -7,8 +7,12 @@ import android.os.Bundle
 
-    override fun onActivityCreated(activity: Activity, bundle: Bundle?) {}
+    override fun onActivityCreated(activity: Activity, bundle: Bundle?) {
+        currentActivity = activity
```
Resumed/paused is a better fit for what you're trying to track than created/destroyed.

For instance, if you go from EmulationActivity to SettingsActivity, EmulationActivity will become paused and SettingsActivity will become created, started, and resumed. Then when returning to EmulationActivity, SettingsActivity will become paused, stopped and destroyed, and EmulationActivity will become resumed. In this scenario, the current logic would cause currentActivity to be set to null instead of the EmulationActivity instance.

JosJuice reviewed (commented):
Sorry for the delay in following up. I haven't been able to review all the USB and audio code in detail, but I've caught a few things.

MayImilae commented:
> > When I tested this I only saw what was in the top window. What scenario does that occur?
> 
> Currently, the microphone configuration groupbox is displayed only when the Emulate Wii Speak checkbox is checked. So the groupbox should (dis)appear when the checkbox state is changed. If there is a bug and it doesn't behave properly, then I'll have to investigate to see what's causing this.

Confirmed that was how it works. Probably just an oversight on my part.

sepalani commented:
I addressed some of the comments. I still need to address the points related to Android, the config name and I will try something else regarding the UI since it might confuse users.

JosJuice commented on Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/features/settings/ui/viewholder/SwitchSettingViewHolder.kt:
```diff
@@ -57,6 +59,13 @@ class SwitchSettingViewHolder(
 
+            if (setting.setting === BooleanSetting.MAIN_EMULATE_WII_SPEAK && isChecked) {
+                if (!PermissionsHandler.hasRecordAudioPermission(itemView.context)) {
+                  val currentActivity = adapter.fragmentActivity as Activity;
```
```suggestion
                  val currentActivity = adapter.fragmentActivity as Activity
```

sepalani commented:
I think I addressed all the remaining comments. Since we can't easily (dis)connect emulated USB devices on the fly, I removed the unused function.

This PR should be ready to be reviewed, tested and merged.

JMC47 commented:
Any last words on this?  I'll do another test through it tomorrow.  Is there are no more reviews it's getting merged.

iwubcode commented on Source/Core/AudioCommon/CubebUtils.cpp:
```diff
@@ -82,3 +91,146 @@ std::shared_ptr<cubeb> CubebUtils::GetContext()
+  return devices;
+}
+
+cubeb_devid GetInputDeviceById(const std::string& id)
```
Can `id` be a `string_view`?

sepalani replied:
Yes, good catch.

iwubcode commented on Source/Core/AudioCommon/CubebUtils.cpp:
```diff
@@ -82,3 +91,146 @@ std::shared_ptr<cubeb> CubebUtils::GetContext()
   return shared;
 }
+
+std::vector<std::pair<std::string, std::string>> ListInputDevices()
```
These strings ultimately come from a cubeb context which is a static.  Are we able to reference these names as a `string_view` instead?

sepalani replied:
I don't think that's safe since these strings come from a collection that I need to destroy at the end of the function. I also couldn't find examples in cubeb source code where they're using `string_view` to reference them.

JMC47 commented:
Last call for reviews on this.  It's been waiting very long time and if I test it and the use cases are working on Linux and Windows, I will be merging it unless someone posts here and says they plan on reviewing it again or tells me not to merge it.

sepalani commented:
I rebased this PR and added some kind of loudness level calculation. I'm not an audio expert so it might be worth checking that the formulas I'm using are correct.

I also switched this PR back to draft as I'm doing some more testing regarding the loudness level calculation.

sepalani commented:
I rebased this PR and added another commit that fixes an oversight regarding the [default sampling rate](https://github.com/dolphin-emu/dolphin/pull/12769/commits/dbfb348a4e12eee600b6146d918271ce76a96ec2#diff-ecb85a064899a5ec8ec30565fa088609a6c3ec462b4d4e8c5949e986569f089fR61) _(I assume it was defaulted to another value [for convenience](https://github.com/dolphin-emu/dolphin/pull/12769/commits/dbfb348a4e12eee600b6146d918271ce76a96ec2#diff-1bdac189795f02d6b7c6f5b5e8b370a3575359980e2cec43f10e955e9a9cd845R104) since changing the sampling rate wasn't implemented)_. This could have caused audio distortions for games not using the 8KHz sampling rate with the previous builds of this PR.

Otherwise, I tested this PR on Windows and Linux with Monster Hunter Tri, Animal Crossing and the Wii Speak channel. I also tested the Wii Speak channel on Android but couldn't for the others as I can't get them to run at a steady full speed. I couldn't find issues while testing.

Depending on how the real microphone is configured, you might want to adjust the microphone volume or the Wii Speak volume in the Wii Speak settings:
![image](https://github.com/user-attachments/assets/c46fcdd6-6ecd-496e-b5d7-8f5d889764a7)

It appears that my laptop microphone is quite sensitive and on MH3, I appear to be chatting constantly (I suppose my microphone might be detecting the fan), adjusting the volume seems to fix the issue.

Ready to be reviewed & merged.

realSquidCoder commented:
Sounds awesome! I'll go ahead and give this some tests on my devices too after i get off work and report any bugs i find

JMC47 commented:
The latest version of this Pull Request functions incredibly well.  I essentially got my Wii Speak games (Wheel of Fortune and Jeopardy) to function better than they do on console once I set the Volume properly (0DB is what it wanted for my microphone.)  

I found the UI to be a bit cumbersome on my machine (the -50db Volume Modifier and +50db is really crunched close making me mistake what it was actually saying!) but overall this is absolutely worth having and a great addition.

sepalani commented:
PR rebased and updated since the USBScanner code addition. I updated the copyright year to 2025 while I was at it. I retested this PR with the games and systems I have and it seems to work like before.

On a side note, I noticed a regression preventing the debugger UI to show up. I had to enable achievements, disable hardcore mode, then disabling achievements again to fix the issue. I suppose there is a missing check somewhere before enforcing hardcore mode.

Regardless, this PR is ready to be reviewed, tested and merged.

JMC47 commented:
Yeah, that bug is known and we're working on fixing the debug menu.

OatmealDome reviewed (changes requested):
This PR fails to build with `-DENABLE_CUBEB=OFF`. I've included the error I received, though after fixing it there are probably more:

```
/repo/Source/Core/Core/IOS/USB/Emulated/Microphone.cpp:8:10: fatal error: 'cubeb/cubeb.h' file not found
    8 | #include <cubeb/cubeb.h>
      |          ^~~~~~~~~~~~~~~
```

jordan-woyak commented on Source/Core/DolphinQt/EmulatedUSB/WiiSpeakWindow.h:
```diff
@@ -0,0 +1,34 @@
+  QCheckBox* m_checkbox_enabled;
+  QComboBox* m_combobox_microphones;
+  QGroupBox* m_config_group;
+  QLabel* m_slider_label;
```
`m_config_group` and `m_slider_label` don't need to be member variables. They are just used in one function.

sepalani commented:
@jordan-woyak @OatmealDome 
Thank you, all your comments should be addressed. I also added a way to build without cubeb on Windows.

jordan-woyak commented on Source/Core/Core/IOS/USB/Emulated/Microphone.h:
```diff
@@ -0,0 +1,131 @@
+    static_assert((SAMPLES_NEEDED % BUFF_SIZE_SAMPLES) == 0);
+
+    static constexpr FloatType MAX_AMPLITUDE =
+        FloatType{std::numeric_limits<UnsignedSampleType>::max() / 2};
```
This doesn't resolve the warning. You're still doing integer division `(65535 / 2)`. Do you want the value to be `32767.5f` or `32767` ?

sepalani replied:
I want an integer division (and its truncation) `32767` to be converted to a float type (here float but could be changed later).

Where can I find the warning?

jordan-woyak replied:
The warning comes from this: https://clang.llvm.org/extra/clang-tidy/checks/bugprone/integer-division.html

sepalani replied:
I had to use an integer cast instead, in order to silence the warning.

OatmealDome commented:
Build still fails with `-DENABLE_CUBEB=OFF`. Same error as before.

sepalani commented:
@jordan-woyak 
The clang-tidy warning should be fixed.

@OatmealDome 
It should be good on Mac/Linux as well now.

OatmealDome reviewed (approved):
Haven't looked at the code, but this PR does now build with `-DENABLE_CUBEB=OFF`.

jordan-woyak commented on Source/Core/Core/IOS/USB/Emulated/Microphone.cpp:
```diff
@@ -0,0 +1,370 @@
+    m_worker.Execute([this] { m_cubeb_ctx.reset(); });
+}
+
+static void state_callback(cubeb_stream* stream, void* user_data, cubeb_state state)
```
`state_callback` -> `StateCallback`

jordan-woyak commented on Source/Core/AudioCommon/CubebUtils.cpp:
```diff
@@ -82,3 +91,146 @@ std::shared_ptr<cubeb> CubebUtils::GetContext()
+                 "\tGroup ID: {}\n"
+                 "\tVendor: {}\n"
+                 "\tState: {}",
+                 i, info.device_id, info.friendly_name, info.group_id,
```
This throws an exception on my system from `info.group_id` being `nullptr`. Maybe all these strings need to be checked like you did with `vendor_name`.

sepalani replied:
I added more sanity checks to prevent it from happening. Sadly, I wasn't expecting a nullptr here as it wasn't mentioned in the struct description in cubeb.h :v/.

jordan-woyak commented on Source/Core/Core/IOS/USB/Emulated/Microphone.cpp:
```diff
@@ -0,0 +1,370 @@
+    u32 minimum_latency;
+    if (cubeb_get_min_latency(m_cubeb_ctx.get(), &params, &minimum_latency) != CUBEB_OK)
+    {
+      WARN_LOG_FMT(IOS_USB, "Error getting minimum latency");
```
`minimum_latency` should probably at least be initialized if this is going to just be a `WARN_LOG`.

jordan-woyak commented on Source/Core/Core/IOS/USB/Emulated/Microphone.h:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    SampleType GetPeak() const;
+    FloatType GetDecibel(FloatType value) const;
```
`GetDecibel` can be `static`.

jordan-woyak commented on Source/Core/Core/IOS/USB/Emulated/Microphone.h:
```diff
@@ -0,0 +1,131 @@
+    FloatType GetRootMeanSquareDb() const;
+    FloatType GetCrestFactor() const;
+    FloatType GetCrestFactorDb() const;
+    FloatType ComputeGain(FloatType db) const;
```
`ComputeGain` can be `static`

jordan-woyak commented on Source/Core/Core/IOS/USB/Emulated/WiiSpeak.h:
```diff
@@ -0,0 +1,101 @@
+{
+public:
+  WiiSpeak();
+  ~WiiSpeak();
```
```suggestion
  ~WiiSpeak() override;
```

jordan-woyak commented on Source/Core/DolphinQt/EmulatedUSB/WiiSpeakWindow.cpp:
```diff
@@ -0,0 +1,149 @@
+  Config::SetBaseOrCurrent(Config::MAIN_EMULATE_WII_SPEAK, emulate);
+}
+
+void WiiSpeakWindow::SetWiiSpeakConnectionState(bool muted)
```
The function is called `SetWiiSpeakConnectionState` but the setting is called `MAIN_WII_SPEAK_MUTED`?

sepalani replied:
This was an oversight when renaming the config from MAIN_WII_SPEAK_CONNECTED to MAIN_WII_SPEAK_MUTED. Should be fixed, now.

sepalani commented:
@jordan-woyak 
All comments should be addressed.

I rebased the PR to solve the merge conflict.

During the rebase, I also had to change CoInitSyncWorker constructor (string_view -> string) in respect to the WorkQueueThread(SP) interface change.

Since I migrated to a new laptop under Windows 11, I was able to test a different microphone.
How much the OS and the driver can have impacts on the voice recording is huge:
 - Using the default device and settings, my voice is barely audible (I had to set the volume modifier to +20dB).
 - Using the driver dedicated hi-res device, my voice was crystal clear without needing to adjust the volume.

In sum, depending on the system settings the OS and/or the driver might:
 - Adjust the microphone volume or apply filters when it detects a voice (compared to other kinds of sound).
 - Adjust the microphone volume or apply filters based on noise canceling and echo reduction technologies.

Having a volume slider seems to be a nice workaround when you don't want to mess with the system or driver settings to adjust the microphone volume.

JMC47 commented:
Yeah, I noticed that too.  Different computers need different volumes so I think having a slider makes sense.

jordan-woyak commented on Source/Core/Core/IOS/USB/Emulated/Microphone.cpp:
```diff
@@ -0,0 +1,371 @@
+  if (Config::Get(Config::MAIN_WII_SPEAK_MUTED))
+    return nframes;
+
+  auto* mic = static_cast<Microphone*>(user_data);
```
This function would be easier to look at if you made `DataCallback` call a non-static member function once you get the `this` pointer from `user_data`, so all the `mic->m_private_member` can just be `m_private_member`.

sepalani replied:
Nice idea. It allowed to make the `UpdateLoudness` method private and remove the `GetSampler` method.

jordan-woyak commented on Source/Core/Core/IOS/USB/Emulated/Microphone.cpp:
```diff
@@ -0,0 +1,371 @@
+                             return MathUtil::SaturatingCast<s16>(sample * gain);
+                           });
+
+  mic->UpdateLoudness(processed_buff_in);
```
It looks like this writes `m_loudness_level` from the cubeb-callback-thread which is read unprotected from core-thread in `GetLoudnessLevel` so I think you need to at least make `m_loudness_level` atomic.

jordan-woyak commented on Source/Core/Core/IOS/USB/Emulated/Microphone.cpp:
```diff
@@ -0,0 +1,371 @@
+  if (mic->m_samples_avail > mic->STREAM_SIZE)
+  {
+    WARN_LOG_FMT(IOS_USB, "Wii Speak ring buffer is full, data will be lost!");
+    mic->m_samples_avail = 0;
```
Similar situation as with `m_loudness_level`. `m_samples_avail` is read from core-thread unprotected, but it's maybe a bit more complicated with the buffer-overflow logic here.

Does setting `m_samples_avail` to zero like this even make sense? Maybe it should just be capped to `STREAM_SIZE` ? I think it can then just be atomic.

sepalani replied:
Not sure how to make it safe with the ring buffer logic so I added a `std::lock_guard` and marked the `std::mutex` as `mutable`.

Regarding `m_samples_avail = 0` I was hoping that it will let the CPU catch up. However, as you pointed out, it doesn't make much sense. If the CPU is catching up, it will catch up with the buffer being full regardless.

sepalani commented:
@jordan-woyak 
I addressed your comments and I also rebased this PR to use the recently merged `AsyncWorkerThread`.

jordan-woyak commented on Source/Core/Core/IOS/USB/Emulated/Microphone.cpp:
```diff
@@ -0,0 +1,371 @@
+
+  UpdateLoudness(processed_buff_in);
+
+  for (s16 le_sample : processed_buff_in)
```
`processed_buff_in` is a `transform_view`. By passing that directly to `UpdateLoudness` you're doing the gain and SaturatingCast calculations every time it gets iterated within `UpdateLoudness -> Loudness::Update` (3 times?). Then you do it again in the for-loop below.

sepalani commented:
@jordan-woyak 
I updated the code so it loops over the view once and the loudness update is done sample by sample.

jordan-woyak commented on Source/Core/Core/IOS/USB/Emulated/Microphone.cpp:
```diff
@@ -0,0 +1,383 @@
+  std::lock_guard lock(m_ring_lock);
+
+  // Skip data if sampling is off or mute is on
+  if (!m_sampler.sample_on || m_sampler.mute)
```
This `m_sampler` object is shared with `WiiSpeak` which writes to it from another thread and doesn't hold this mutex.

It looks like you can just make those members atomic.

jordan-woyak commented on Source/Core/Core/IOS/USB/Emulated/Microphone.cpp:
I know it's a debug log, but one every 128 samples is maybe excessive? Maybe just rely on the `WII_SPEAK_LOG_STATS`macro  if someone wants this logged?

sepalani replied:
That's a valid concern and can bother someone checking other IOS_USB debug logs so I'll rely on the macro as you suggested.

sepalani commented:
@jordan-woyak 
I rebased this PR and as suggested I squashed together similar commits.
